File: Scripts/audit_dense_bigrams.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Independent, read-only Ruby attestation of the dense German-bigram receipt.
# This implementation consumes only the counted fixture and frozen P1030684
# plaintext. It deliberately does not load HELUT, Swift output, or Python output.

require "digest"
require "json"

ALPHABET_WIDTH = 26
SMOOTHING_MASS = 0.5
FIXTURE_DIGEST = "06f8694ec17906e993223e0179c6ca326030e62f77ec86e5c4120fddcf68f111"
LETTER_TOTAL = 28_508_834
OBSERVED_TOTAL = 674
ABSENT_GRAMS = ["JX", "QY"].freeze
PLAINTEXT_DIGEST = "e355b691e78cd71bc1ed816e0808ab4f18aeb73376008b2a71dced197ad07145"
P1030684_PLAINTEXT = [
  "VVVUUUVIRSOBENNULEINSXXMITUUUVIRSIBENNULZWOYVIRSIBENNULDREIY",
  "ZWODREISECHSEINSYZWODREIDREIACHTEINSDREIOITNACHWZSTENPASSIRT"
].join.freeze

class ReceiptFailure < StandardError; end

def assert_receipt(condition, message)
  raise ReceiptFailure, message unless condition
end

def numeric_double_bits(number)
  [number].pack("G").unpack("Q>").fetch(0)
end

def parse_counted_fixture(filename)
  bytes = File.binread(filename)
  actual_digest = Digest::SHA256.hexdigest(bytes)
  assert_receipt(
    actual_digest == FIXTURE_DIGEST,
    "fixture SHA-256 changed: expected #{FIXTURE_DIGEST}, got #{actual_digest}"
  )

  source = bytes.dup.force_encoding(Encoding::US_ASCII)
  assert_receipt(source.valid_encoding?, "fixture is not strict ASCII")
  records = source.lines.map(&:chomp)
  assert_receipt(!records.empty?, "fixture is empty")

  metadata = /\A# letters=(\d+) ic=([0-9]+(?:\.[0-9]+)?) distinct=(\d+)\z/.match(records.shift)
  assert_receipt(!metadata.nil?, "fixture metadata header is malformed")
  letters = Integer(metadata[1], 10)
  ic_string = metadata[2]
  declared_entries = Integer(metadata[3], 10)

  by_gram = {}
  encountered_order = []
  records.each_with_index do |record, offset|
    parsed = /\A([A-Z]{2}) ([0-9]+)\z/.match(record)
    assert_receipt(!parsed.nil?, "malformed fixture line #{offset + 2}: #{record.inspect}")
    gram = parsed[1]
    assert_receipt(!by_gram.key?(gram), "duplicate gram #{gram} at fixture line #{offset + 2}")
    count = Integer(parsed[2], 10)
    assert_receipt(count.positive?, "non-positive observed count for #{gram}")
    assert_receipt(count <= 0xffff_ffff, "count for #{gram} exceeds Swift UInt32")
    by_gram[gram] = count
    encountered_order << gram
  end

  assert_receipt(encountered_order == encountered_order.sort, "fixture entries are not lexical")
  assert_receipt(letters == LETTER_TOTAL, "unexpected letters metadata: #{letters}")
  assert_receipt(declared_entries == OBSERVED_TOTAL, "unexpected distinct metadata: #{declared_entries}")
  assert_receipt(by_gram.length == declared_entries, "declared and parsed entry counts differ")

  universe = []
  ALPHABET_WIDTH.times do |first|
    ALPHABET_WIDTH.times do |second|
      universe << (65 + first).chr + (65 + second).chr
    end
  end
  absent = universe.reject { |gram| by_gram.key?(gram) }
  assert_receipt(absent == ABSENT_GRAMS, "unexpected missing grams: #{absent.inspect}")

  pair_total = 0
  by_gram.each_value { |count| pair_total += count }
  assert_receipt(pair_total == letters - 1, "pair total is not letters - 1")

  cells = Array.new(ALPHABET_WIDTH * ALPHABET_WIDTH, 0)
  by_gram.each do |gram, count|
    row = gram.getbyte(0) - 65
    column = gram.getbyte(1) - 65
    cells[row * ALPHABET_WIDTH + column] = count
  end

  details = {
    "sha256" => actual_digest,
    "letters" => letters,
    "ic" => ic_string,
    "declared_distinct" => declared_entries,
    "parsed_distinct" => by_gram.length,
    "counted_pairs" => pair_total,
    "missing" => absent
  }
  [cells, details]
end

def conditional_logs(cells)
  probabilities = Array.new(cells.length, 0.0)
  row = 0
  while row < ALPHABET_WIDTH
    offset = row * ALPHABET_WIDTH
    row_count = 0
    column = 0
    while column < ALPHABET_WIDTH
      row_count += cells[offset + column]
      column += 1
    end

    denominator = row_count.to_f + SMOOTHING_MASS * ALPHABET_WIDTH.to_f
    column = 0
    while column < ALPHABET_WIDTH
      numerator = cells[offset + column].to_f + SMOOTHING_MASS
      probabilities[offset + column] = Math.log(numerator / denominator)
      column += 1
    end
    row += 1
  end
  probabilities
end

def fold_control(cells, probabilities)
  digest = Digest::SHA256.hexdigest(P1030684_PLAINTEXT)
  assert_receipt(digest == PLAINTEXT_DIGEST, "control plaintext digest changed: #{digest}")
  assert_receipt(P1030684_PLAINTEXT.bytesize == 120, "control plaintext is not 120 symbols")

  total = 0.0
  floor_starts = []
  first_index = nil
  last_index = nil
  cursor = 0
  while cursor < P1030684_PLAINTEXT.bytesize - 1
    left = P1030684_PLAINTEXT.getbyte(cursor) - 65
    right = P1030684_PLAINTEXT.getbyte(cursor + 1) - 65
    assert_receipt(left.between?(0, 25) && right.between?(0, 25), "control contains non-A-Z data")
    index = left * ALPHABET_WIDTH + right
    first_index = index if first_index.nil?
    last_index = index
    floor_starts << cursor if cells[index].zero?
    total = total + probabilities[index]
    cursor += 1
  end

  assert_receipt(cursor == 119, "control did not produce 119 windows")
  mean = total / cursor.to_f
  control = {
    "plaintext_sha256" => digest,
    "symbols" => P1030684_PLAINTEXT.bytesize,
    "windows" => cursor,
    "first_index" => first_index,
    "last_index" => last_index,
    "floor_window_starts" => floor_starts
  }
  [control, total, mean]
end

def run_audit
  raise ReceiptFailure, "usage: #{File.basename($PROGRAM_NAME)} [fixture]" if ARGV.length > 1
  default_fixture = File.expand_path("../Fixtures/german_bigrams.txt", __dir__)
  fixture_name = File.expand_path(ARGV.fetch(0, default_fixture))

  cells, fixture = parse_counted_fixture(fixture_name)
  control, total, mean = fold_control(cells, conditional_logs(cells))
  total_bits = numeric_double_bits(total)
  mean_bits = numeric_double_bits(mean)

  {
    "schema" => "helut.dense-bigram-audit.v1",
    "implementation" => "ruby-independent-v1",
    "runtime" => {
      "ruby" => RUBY_DESCRIPTION,
      "platform" => RUBY_PLATFORM
    },
    "fixture" => fixture,
    "control" => control,
    "scoring" => {
      "layout" => "row-major A-Z conditional bigrams",
      "add_k" => SMOOTHING_MASS,
      "logarithm" => "natural",
      "accumulation" => "explicit left fold from +0.0",
      "mean_denominator" => control.fetch("windows")
    },
    "result" => {
      "total_decimal" => format("%.17g", total),
      "mean_decimal" => format("%.17g", mean),
      "total_bit_pattern" => format("0x%016x", total_bits),
      "mean_bit_pattern" => format("0x%016x", mean_bits)
    }
  }
end

begin
  puts JSON.pretty_generate(run_audit)
rescue ReceiptFailure, SystemCallError => error
  warn "dense-bigram audit failed: #{error.message}"
  exit 1
end

File: Scripts/audit_dense_bigrams.py
#!/usr/bin/env python3
"""Independently attest the dense German-bigram P1030684 receipt.

This read-only program intentionally does not import HELUT code or generated Swift
constants. It parses the committed counted-bigram fixture, reconstructs the
row-normalized add-0.5 model, and scores the frozen 120-symbol control with an
explicit binary64 left fold. The numeric IEEE-754 patterns are written to stdout
for comparison with a separately implemented Ruby audit and Swift under test.
"""

import argparse
import hashlib
import json
import math
import platform
import re
import struct
import sys
from pathlib import Path

ALPHABET_SIZE = 26
ADD_K = 0.5
EXPECTED_FIXTURE_SHA256 = (
    "06f8694ec17906e993223e0179c6ca326030e62f77ec86e5c4120fddcf68f111"
)
EXPECTED_LETTERS = 28_508_834
EXPECTED_DISTINCT = 674
EXPECTED_MISSING = ("JX", "QY")
EXPECTED_PLAINTEXT_SHA256 = (
    "e355b691e78cd71bc1ed816e0808ab4f18aeb73376008b2a71dced197ad07145"
)
CONTROL_PLAINTEXT = (
    "VVVUUUVIRSOBENNULEINSXXMITUUUVIRSIBENNULZWOYVIRSIBENNULDREIY"
    "ZWODREISECHSEINSYZWODREIDREIACHTEINSDREIOITNACHWZSTENPASSIRT"
)
HEADER_PATTERN = re.compile(
    r"# letters=(\d+) ic=([0-9]+(?:\.[0-9]+)?) distinct=(\d+)"
)
ENTRY_PATTERN = re.compile(r"([A-Z]{2}) ([0-9]+)")


class AuditError(RuntimeError):
    """The frozen audit input or protocol did not satisfy an invariant."""


def require(condition: bool, message: str) -> None:
    if not condition:
        raise AuditError(message)


def binary64_bits(value: float) -> str:
    numeric_bits = struct.unpack(">Q", struct.pack(">d", value))[0]
    return f"0x{numeric_bits:016x}"


def parse_fixture(path: Path):
    raw = path.read_bytes()
    digest = hashlib.sha256(raw).hexdigest()
    require(
        digest == EXPECTED_FIXTURE_SHA256,
        f"fixture SHA-256 changed: expected {EXPECTED_FIXTURE_SHA256}, got {digest}",
    )

    try:
        text = raw.decode("ascii")
    except UnicodeDecodeError as error:
        raise AuditError("fixture is not strict ASCII") from error

    lines = text.splitlines()
    require(bool(lines), "fixture is empty")
    header_match = HEADER_PATTERN.fullmatch(lines[0])
    require(header_match is not None, "fixture metadata header is malformed")
    letters = int(header_match.group(1))
    ic_text = header_match.group(2)
    declared_distinct = int(header_match.group(3))

    entries = {}
    lexical_order = []
    for line_number, line in enumerate(lines[1:], start=2):
        require(bool(line), f"blank line at fixture line {line_number}")
        match = ENTRY_PATTERN.fullmatch(line)
        require(match is not None, f"malformed fixture line {line_number}: {line!r}")
        gram, count_text = match.groups()
        require(gram not in entries, f"duplicate gram {gram} at fixture line {line_number}")
        count = int(count_text)
        require(count > 0, f"non-positive observed count for {gram}")
        require(count <= 0xFFFF_FFFF, f"count for {gram} exceeds Swift UInt32")
        entries[gram] = count
        lexical_order.append(gram)

    require(lexical_order == sorted(lexical_order), "fixture entries are not lexical")
    require(letters == EXPECTED_LETTERS, f"unexpected letters metadata: {letters}")
    require(
        declared_distinct == EXPECTED_DISTINCT,
        f"unexpected distinct metadata: {declared_distinct}",
    )
    require(
        len(entries) == declared_distinct,
        f"metadata says {declared_distinct} entries but parsed {len(entries)}",
    )

    expected_grams = {
        chr(65 + first) + chr(65 + second)
        for first in range(ALPHABET_SIZE)
        for second in range(ALPHABET_SIZE)
    }
    missing = tuple(sorted(expected_grams.difference(entries)))
    require(missing == EXPECTED_MISSING, f"unexpected missing grams: {missing}")

    counted_pairs = 0
    for count in entries.values():
        counted_pairs += count
    require(
        counted_pairs == letters - 1,
        f"pair total {counted_pairs} does not equal letters - 1 ({letters - 1})",
    )

    counts = [0] * (ALPHABET_SIZE * ALPHABET_SIZE)
    for gram, count in entries.items():
        first = ord(gram[0]) - 65
        second = ord(gram[1]) - 65
        counts[first * ALPHABET_SIZE + second] = count

    return counts, {
        "sha256": digest,
        "letters": letters,
        "ic": ic_text,
        "declared_distinct": declared_distinct,
        "parsed_distinct": len(entries),
        "counted_pairs": counted_pairs,
        "missing": list(missing),
    }


def build_log_probabilities(counts):
    log_probabilities = [0.0] * len(counts)
    for first in range(ALPHABET_SIZE):
        row_total = 0
        row_offset = first * ALPHABET_SIZE
        for second in range(ALPHABET_SIZE):
            row_total += counts[row_offset + second]
        denominator = float(row_total) + ADD_K * float(ALPHABET_SIZE)
        for second in range(ALPHABET_SIZE):
            numerator = float(counts[row_offset + second]) + ADD_K
            log_probabilities[row_offset + second] = math.log(numerator / denominator)
    return log_probabilities


def score_control(counts, log_probabilities):
    plaintext_bytes = CONTROL_PLAINTEXT.encode("ascii")
    plaintext_digest = hashlib.sha256(plaintext_bytes).hexdigest()
    require(
        plaintext_digest == EXPECTED_PLAINTEXT_SHA256,
        f"control plaintext changed: {plaintext_digest}",
    )
    require(len(plaintext_bytes) == 120, f"control length is {len(plaintext_bytes)}, not 120")

    total = 0.0
    indices = []
    floor_window_starts = []
    for start in range(len(plaintext_bytes) - 1):
        first = plaintext_bytes[start] - 65
        second = plaintext_bytes[start + 1] - 65
        require(0 <= first < 26 and 0 <= second < 26, "control contains a non-A-Z byte")
        table_index = first * ALPHABET_SIZE + second
        indices.append(table_index)
        if counts[table_index] == 0:
            floor_window_starts.append(start)
        total = total + log_probabilities[table_index]

    windows = len(indices)
    require(windows == 119, f"control produced {windows} windows, not 119")
    mean = total / float(windows)
    return {
        "plaintext_sha256": plaintext_digest,
        "symbols": len(plaintext_bytes),
        "windows": windows,
        "first_index": indices[0],
        "last_index": indices[-1],
        "floor_window_starts": floor_window_starts,
    }, total, mean


def main() -> int:
    root = Path(__file__).resolve().parent.parent
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--fixture",
        type=Path,
        default=root / "Fixtures" / "german_bigrams.txt",
        help="counted bigram fixture to attest (default: repository dense fixture)",
    )
    args = parser.parse_args()

    counts, fixture_receipt = parse_fixture(args.fixture.resolve())
    log_probabilities = build_log_probabilities(counts)
    control_receipt, total, mean = score_control(counts, log_probabilities)

    receipt = {
        "schema": "helut.dense-bigram-audit.v1",
        "implementation": "python-independent-v1",
        "runtime": {
            "python": sys.version.splitlines()[0],
            "platform": platform.platform(),
            "architecture": platform.machine(),
        },
        "fixture": fixture_receipt,
        "control": control_receipt,
        "scoring": {
            "layout": "row-major A-Z conditional bigrams",
            "add_k": ADD_K,
            "logarithm": "natural",
            "accumulation": "explicit left fold from +0.0",
            "mean_denominator": control_receipt["windows"],
        },
        "result": {
            "total_decimal": repr(total),
            "mean_decimal": repr(mean),
            "total_bit_pattern": binary64_bits(total),
            "mean_bit_pattern": binary64_bits(mean),
        },
    }
    print(json.dumps(receipt, indent=2, sort_keys=True))
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except (AuditError, OSError) as error:
        print(f"dense-bigram audit failed: {error}", file=sys.stderr)
        raise SystemExit(1)
